test/e2e/cluster_authorized_cidrs_connectivity.go

```diff
@@ -0,0 +1,255 @@
+package e2e
+
+import (
+	"context"
+	"crypto/tls"
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/to"
+
+	"github.com/Azure/ARO-HCP/test/util/framework"
+	"github.com/Azure/ARO-HCP/test/util/labels"
+)
+
+var _ = Describe("Authorized CIDRs", func() {
+	Context("Connectivity", func() {
+		It("should allow API access only from authorized VM IP",
+			labels.RequireNothing,
+			labels.Critical,
+			labels.Positive,
+			func(ctx context.Context) {
+				const (
+					clusterName                      = "cidr-connectivity-test"
+					customerNetworkSecurityGroupName = "customer-nsg-name"
+					customerVnetName                 = "customer-vnet-name"
+					customerVnetSubnetName           = "customer-vnet-subnet1"
+					openshiftControlPlaneVersionId   = "4.19"
+				)
+
+				tc := framework.NewTestContext()
+
+				By("creating a resource group")
+				resourceGroup, err := tc.NewResourceGroup(ctx, "e2e-cidr-connectivity", tc.Location())
+				Expect(err).NotTo(HaveOccurred())
+
+				By("creating cluster parameters")
+				clusterParams := framework.NewDefaultClusterParams()
+				clusterParams.ClusterName = clusterName
+				managedResourceGroupName := framework.SuffixName(*resourceGroup.Name, "-managed", 64)
+				clusterParams.ManagedResourceGroupName = managedResourceGroupName
+				clusterParams.OpenshiftVersionId = openshiftControlPlaneVersionId
+
+				By("creating customer resources")
+				clusterParams, err = tc.CreateClusterCustomerResources(ctx,
+					resourceGroup,
+					clusterParams,
+					map[string]interface{}{
+						"persistTagValue":        false,
+						"customerNsgName":        customerNetworkSecurityGroupName,
+						"customerVnetName":       customerVnetName,
+						"customerVnetSubnetName": customerVnetSubnetName,
+					},
+					TestArtifactsFS,
+				)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("generating SSH key pair for VM")
+				sshPublicKey, _, err := framework.GenerateSSHKeyPair()
+				Expect(err).NotTo(HaveOccurred())
+
+				By("deploying test VM")
+				vmName := fmt.Sprintf("%s-test-vm", clusterName)
+				vmDeployment, err := tc.CreateBicepTemplateAndWait(ctx,
+					*resourceGroup.Name,
+					"test-vm",
+					framework.Must(TestArtifactsFS.ReadFile("test-artifacts/generated-test-artifacts/modules/test-vm.json")),
+					map[string]interface{}{
+						"vmName":       vmName,
+						"vnetName":     customerVnetName,
+						"subnetName":   customerVnetSubnetName,
+						"sshPublicKey": sshPublicKey,
+					},
+					30*time.Minute,
+				)
+				Expect(err).NotTo(HaveOccurred())
+
+				By("extracting VM public IP from deployment outputs")
+				vmPublicIP, err := framework.GetOutputValueString(vmDeployment, "publicIP")
+				Expect(err).NotTo(HaveOccurred())
+				Expect(vmPublicIP).NotTo(BeEmpty(), "VM public IP should be in deployment outputs")
+
+				By("setting authorized CIDRs with VM IP")
```

Comment:
Any way to not make this dependent on the VM already running? Does it need a public IP? Any network magic to use something internal to this resource group?

Reply:
That wont work since we are pinging the api server the request passes through the internet. If the traffic were to be strictly inside the vnet or subnet than once can consider some network magic.

What is tested here is the likely scenario the user would do.